File: 2022/CPP/02/main.cpp
#include <iostream>
#include <fstream>
#include <vector>

std::vector<std::string> split(std::string line, char delimiter);


int main() {
    std::ifstream input("guide.txt");
    int total = 0;

    for (std::string line; getline(input, line);) {
        std::vector<std::string> splitLine = split(line, ' ');

        std::string opponentChoose = splitLine[0];
        std::string issue = splitLine[1];

        if (issue == "X") {
            issue = "loose";
            total += 0;
        } else if (issue == "Y") {
            issue = "tie";
            total += 3;
        } else if (issue == "Z") {
            issue = "win";
            total += 6;
        }

        /* Reminds:
        rock = 1
        paper = 2
        scissors = 3
        */

        if (opponentChoose == "A") { // rock
            if (issue == "win") {
                total += 2;
            } else if (issue == "tie") {
                total += 1;
            } else if (issue == "loose") {
                total += 3;
            }
        } else if (opponentChoose == "B") { // paper
            if (issue == "win") {
                total += 3;
            } else if (issue == "tie") {
                total += 2;
            } else if (issue == "loose") {
                total += 1;
            }
        } else if (opponentChoose == "C") { // Scissors
            if (issue == "win") {
                total += 1;
            } else if (issue == "tie") {
                total += 3;
            } else if (issue == "loose") {
                total += 2;
            }
        }
    }

    std::cout << "Total: " << total << std::endl;

    return 0;
}

std::vector<std::string> split(std::string line, char delimiter) {
    std::vector<std::string> splitLine;
    std::string word = "";

    for (char c : line) {
        if (c == delimiter) {
            splitLine.push_back(word);
            word = "";
        } else {
            word += c;
        }
    }

    splitLine.push_back(word);

    return splitLine;
}

File: 2022/Kotlin/02/main.kt
import java.io.File

val rock: String = "A"
val paper: String = "B"
val scissors: String = "C"

val win: String = "Z"
val tie: String = "Y"
val loose: String = "X"

fun main() {
    val guide = File("guide.txt")
    val guideLines = guide.readLines()
    var total = 0

    guideLines.forEach {
        if (it == "") return

        val opponentChoose = it.split(" ")[0]
        var issue = it.split(" ")[1]
        var points = 0

        when {
            issue == "X" -> {
                issue = loose
                points += 0
            }
            issue == "Y" -> {
                issue = tie
                points += 3
            }
            issue == "Z" -> {
                issue = win
                points += 6
            }
        }

        val moove = getChooseTo(opponentChoose, issue)

        when (moove) {
            rock -> points += 1
            paper -> points += 2
            scissors -> points += 3
        }

        total += points
    }

    println("Total points: $total")
}

fun getChooseTo(opponentChoose: String, result: String): String {
    return when (opponentChoose) {
        rock -> {
            when (result) {
                win -> paper
                tie -> rock
                loose -> scissors
                else -> ""
            }
        }
        paper -> {
            when (result) {
                win -> scissors
                tie -> paper
                loose -> rock
                else -> ""
            }
        }
        scissors -> {
            when (result) {
                win -> rock
                tie -> scissors
                loose -> paper
                else -> ""
            }
        }
        else -> ""
    }
}
